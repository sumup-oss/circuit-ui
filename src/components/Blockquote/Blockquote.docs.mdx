import { Status, Props, Story } from '../../../.storybook/components';
import Blockquote from './Blockquote';

# Blockquote

<Status.Stable />

Blockquotes are a way of styling specific text with indention, normally used for demonstrating and highlighting textual information that comes from an external source.

<Story id="components-blockquote--base" />

<Props of={Blockquote} />

## When to use it

For highlighting pieces of textual information, usually quotes, that aren't from SumUp directly.
For instance, we should use a blockquote to highlight a quote from a user that was interviewed on a blog post regarding product development.

## Usage guidelines

- **Do** use blockquotes for citing quotes from users, partners or other external sources
- **Do not** place an image within a blockquote, they should contain only text.

## Component variations

You can customize the size of the blockquote using the `size` property.
We reccomend always sticking to mega in most cases, since it fits well with the Circuit's default body text size.

<Story id="components-blockquote--size" />
