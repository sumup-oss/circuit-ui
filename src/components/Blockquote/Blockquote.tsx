/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { css } from '@emotion/core';

import styled, { StyleProps } from '../../styles/styled';
import Text from '../Text';
import { TextProps } from '../Text/Text';

type Size = 'kilo' | 'mega' | 'giga';

export interface BlockquoteProps extends TextProps {
  /**
   * A Circuit UI body text size.
   */
  size?: Size;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: React.Ref<HTMLQuoteElement>;
}

const baseStyles = ({ theme }: StyleProps) => css`
  label: blockquote;
  padding-left: ${theme.spacings.kilo};
  border-left: 2px solid ${theme.colors.p500};
`;

const gigaStyles = ({ theme, size = 'mega' }: StyleProps & BlockquoteProps) =>
  size === 'giga' &&
  css`
    label: blockquote--giga;
    padding-left: ${theme.spacings.mega};
    border-left: 3px solid ${theme.colors.p500};
  `;

const StyledText = styled(Text)<BlockquoteProps>(baseStyles, gigaStyles);

export function BlockquoteComponent(
  props: BlockquoteProps,
  ref: BlockquoteProps['ref']
) {
  return <StyledText {...props} as="blockquote" italic ref={ref} />;
}

/**
 * Indented and italicised text to denote a quotation.
 */
export const Blockquote = React.forwardRef(BlockquoteComponent);
