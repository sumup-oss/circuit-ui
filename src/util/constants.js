/**
 * Position
 */

export const TOP = 'top';
export const BOTTOM = 'bottom';
export const LEFT = 'left';
export const RIGHT = 'right';

/**
 * Alignment
 */

export const START = 'start';
export const END = 'end';
export const CENTER = 'center';

/**
 * Sizes
 */

export const BIT = 'bit';
export const BYTE = 'byte';
export const KILO = 'kilo';
export const MEGA = 'mega';
export const GIGA = 'giga';
export const TERA = 'tera';
export const PETA = 'peta';
export const EXA = 'exa';
export const ZETTA = 'zetta';

/**
 * Shadows
 */

export const SINGLE = 'single';
export const DOUBLE = 'double';
export const TRIPLE = 'triple';

/**
 * Colors?
 */

export const SUCCESS = 'success';
export const DANGER = 'danger';
export const WARNING = 'warning';
