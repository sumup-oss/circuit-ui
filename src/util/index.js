import * as id from './id';
import * as numbers from './numbers';
import * as currency from './currency';
import * as sharedPropTypes from './shared-prop-types';
import * as sizes from './sizes';
import * as typeCheck from './type-check';

export { id, numbers, currency, sharedPropTypes, sizes, typeCheck };
