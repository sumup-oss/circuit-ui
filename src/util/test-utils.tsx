/**
 * Copyright 2020, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { FunctionComponent } from 'react';
import { renderToStaticMarkup } from 'react-dom/server';
import '@testing-library/jest-dom/extend-expect';
import { configureAxe } from 'jest-axe';
import { render as renderTest, wait, act } from '@testing-library/react';
import { renderHook, act as actHook } from '@testing-library/react-hooks';
import userEvent from '@testing-library/user-event';
import { ThemeProvider } from 'emotion-theming';
import { light } from '@sumup/design-tokens';

import ComponentsContext, {
  defaultComponents,
} from '../components/ComponentsContext';

export type RenderFn = (component: React.ReactElement, ...rest: any) => any;

const WithProviders: FunctionComponent = ({ children }) => (
  <ComponentsContext.Provider value={defaultComponents}>
    <ThemeProvider theme={light}>{children}</ThemeProvider>
  </ComponentsContext.Provider>
);

const renderWithProviders = (renderer: RenderFn): RenderFn => (
  component,
  ...rest
): RenderFn => renderer(<WithProviders>{component}</WithProviders>, rest);

const render: RenderFn = (component, options) =>
  renderTest(component, { wrapper: WithProviders, ...options });
const renderToHtml: RenderFn = renderWithProviders(renderToStaticMarkup);
const create: RenderFn = (...args) => {
  const { container } = render(...args);
  return container.children.length > 1
    ? container.children
    : container.firstChild;
};

const axe = configureAxe({
  rules: {
    // disabled landmark rules when testing isolated components.
    region: { enabled: false },
  },
});

export {
  create,
  render,
  renderToHtml,
  renderHook,
  act,
  actHook,
  userEvent,
  wait,
  axe,
};
