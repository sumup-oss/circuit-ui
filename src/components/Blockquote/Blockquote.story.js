/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import { select, text } from '@storybook/addon-knobs/react';

import docs from './Blockquote.docs.mdx';
import Blockquote from './Blockquote';

const defaultQuote = `
Lorem ipsum dolor amet echo park activated charcoal banjo deep
crucifix pinterest yr af tumeric literally. Tbh four loko tattooed
kickstarter artisan. Lumbersexual tote bag selfies truffaut, tofu vape
tbh adaptogen green juice lo-fi kombucha.
`;

const sizes = [Blockquote.KILO, Blockquote.MEGA, Blockquote.GIGA];

export default {
  title: 'Components|Blockquote',
  component: Blockquote,
  parameters: {
    docs: { page: docs },
    jest: ['Blockquote']
  }
};

export const base = () => (
  <Blockquote size={select('Size', sizes, sizes[0])}>
    {text('Quote', defaultQuote)}
  </Blockquote>
);

export const size = () => (
  <Fragment>
    <Blockquote size={Blockquote.KILO}>
      Kilo - The ability to accept credit card payments that are EMV-compliant
      is essentially an insurance policy against fraud and an impressively
      economical one at that.
    </Blockquote>
    <Blockquote size={Blockquote.MEGA}>
      Mega - The ability to accept credit card payments that are EMV-compliant
      is essentially an insurance policy against fraud and an impressively
      economical one at that.
    </Blockquote>
    <Blockquote size={Blockquote.GIGA}>
      Giga - The ability to accept credit card payments that are EMV-compliant
      is essentially an insurance policy against fraud and an impressively
      economical one at that.
    </Blockquote>
  </Fragment>
);
