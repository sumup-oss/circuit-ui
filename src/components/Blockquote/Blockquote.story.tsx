/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { Fragment } from 'react';
import { select, text } from '@storybook/addon-knobs';

import docs from './Blockquote.docs.mdx';
import { Blockquote } from './Blockquote';

const defaultQuote = `The ability to accept credit card payments that are EMV-compliant is essentially an insurance policy against fraud and an impressively economical one at that.`;

const sizes = ['kilo', 'mega', 'giga'] as const;

export default {
  title: 'Typography/Blockquote',
  component: Blockquote,
  parameters: {
    docs: { page: docs },
    jest: ['Blockquote']
  }
};

export const base = () => (
  <Blockquote size={select('Size', sizes, sizes[1])}>
    {text('Quote', defaultQuote)}
  </Blockquote>
);

export const size = () => (
  <Fragment>
    <Blockquote size="kilo">Kilo - {text('Quote', defaultQuote)}</Blockquote>
    <Blockquote size="mega">Mega - {text('Quote', defaultQuote)}</Blockquote>
    <Blockquote size="giga">Giga - {text('Quote', defaultQuote)}</Blockquote>
  </Fragment>
);
