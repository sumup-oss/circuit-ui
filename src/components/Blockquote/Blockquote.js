/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import { css } from '@emotion/core';

import { childrenPropType } from '../../util/shared-prop-types';
import Text from '../Text';

const baseStyles = ({ theme }) => css`
  label: blockquote;
  padding-left: ${theme.spacings.kilo};
  border-left: 2px solid ${theme.colors.p500};
`;

const gigaStyles = ({ theme, size }) =>
  size === 'giga' &&
  css`
    label: blockquote--giga;
    padding-left: ${theme.spacings.mega};
    border-left: 3px solid ${theme.colors.p500};
  `;

const StyledText = styled(Text)`
  ${baseStyles};
  ${gigaStyles};
`;

/**
 * Indented and italicised text to denote a quotation.
 */
const Blockquote = ({ children, ...props }) => (
  <StyledText {...props} as="blockquote" italic>
    {children}
  </StyledText>
);

// Satisfy react docgen
// const Blockquote = props => <Quote {...props} />;

Blockquote.KILO = 'kilo';
Blockquote.MEGA = 'mega';
Blockquote.GIGA = 'giga';

Blockquote.propTypes = {
  /**
   * Child nodes to be rendered.
   */
  children: childrenPropType,
  /**
   * A Circuit UI body text size.
   */
  size: PropTypes.oneOf([Blockquote.KILO, Blockquote.MEGA, Blockquote.GIGA])
    .isRequired
};

Blockquote.defaultProps = {
  size: Blockquote.KILO
};

/**
 * @component
 */
export default Blockquote;
