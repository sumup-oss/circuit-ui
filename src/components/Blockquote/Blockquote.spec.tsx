/**
 * Copyright 2019, SumUp Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';

import { create, render, renderToHtml, axe } from '../../util/test-utils';

import { Blockquote } from './Blockquote';

describe('Blockquote', () => {
  const quote = `The ability to accept credit card payments that are EMV-compliant is essentially an insurance policy against fraud and an impressively economical one at that.`;

  /**
   * Style tests.
   */
  it('should render with default styles', () => {
    const actual = create(<Blockquote>{quote}</Blockquote>);
    expect(actual).toMatchSnapshot();
  });

  it('should render with kilo styles', () => {
    const actual = create(<Blockquote size="kilo">{quote}</Blockquote>);
    expect(actual).toMatchSnapshot();
  });

  it('should render with giga styles', () => {
    const actual = create(<Blockquote size="giga">{quote}</Blockquote>);
    expect(actual).toMatchSnapshot();
  });

  describe('business logic', () => {
    /**
     * Should accept a working ref
     */
    it('should accept a working ref', () => {
      const tref = React.createRef<HTMLQuoteElement>();
      const { container } = render(
        <Blockquote size="giga" ref={tref}>
          {quote}
        </Blockquote>
      );
      const blockquote = container.querySelector('blockquote');
      expect(tref.current).toBe(blockquote);
    });
  });

  /**
   * Accessibility tests.
   */
  it('should meet accessibility guidelines', async () => {
    const wrapper = renderToHtml(<Blockquote>{quote}</Blockquote>);
    const actual = await axe(wrapper);
    expect(actual).toHaveNoViolations();
  });
});
