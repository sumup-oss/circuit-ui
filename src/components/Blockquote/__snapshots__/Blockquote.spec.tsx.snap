// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Blockquote should render with default styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 16px;
  line-height: 24px;
  font-style: italic;
  padding-left: 12px;
  border-left: 2px solid #3388FF;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 16px;
    line-height: 24px;
  }
}

<blockquote
  class="circuit-0"
>
  The ability to accept credit card payments that are EMV-compliant is essentially an insurance policy against fraud and an impressively economical one at that.
</blockquote>
`;

exports[`Blockquote should render with giga styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 16px;
  line-height: 24px;
  font-style: italic;
  padding-left: 12px;
  border-left: 2px solid #3388FF;
  padding-left: 16px;
  border-left: 3px solid #3388FF;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 18px;
    line-height: 28px;
  }
}

<blockquote
  class="circuit-0"
>
  The ability to accept credit card payments that are EMV-compliant is essentially an insurance policy against fraud and an impressively economical one at that.
</blockquote>
`;

exports[`Blockquote should render with kilo styles 1`] = `
.circuit-0 {
  font-weight: 400;
  margin-bottom: 16px;
  font-size: 13px;
  line-height: 20px;
  font-style: italic;
  padding-left: 12px;
  border-left: 2px solid #3388FF;
}

@media (min-width:480px) {
  .circuit-0 {
    font-size: 13px;
    line-height: 20px;
  }
}

<blockquote
  class="circuit-0"
>
  The ability to accept credit card payments that are EMV-compliant is essentially an insurance policy against fraud and an impressively economical one at that.
</blockquote>
`;
